python/dgl/sampling/labor.py

```diff
@@ -0,0 +1,276 @@
+
+    >>> sg = dgl.sampling.sample_labors(g, [0, 1], 1)
+    >>> sg.edges(order='eid')
+    (tensor([1, 0]), tensor([0, 1]))
+    >>> sg.edata[dgl.EID]
+    tensor([2, 0])
+
+    To sample one inbound edge for node 0 and node 1 with probability in edge feature
+    ``prob``:
+
+    >>> sg = dgl.sampling.sample_labors(g, [0, 1], 1, prob='prob')
+    >>> sg.edges(order='eid')
+    (tensor([2, 1]), tensor([0, 1]))
+
+    With ``fanout`` greater than the number of actual neighbors and without replacement,
+    DGL will take all neighbors instead:
+
+    >>> sg = dgl.sampling.sample_labors(g, [0, 1], 3)
+    >>> sg.edges(order='eid')
+    (tensor([1, 2, 0, 1]), tensor([0, 0, 1, 1]))
+
+    To exclude certain EID's during sampling for the seed nodes:
+
+    >>> g = dgl.graph(([0, 0, 1, 1, 2, 2], [1, 2, 0, 1, 2, 0]))
+    >>> g_edges = g.all_edges(form='all')``
+    (tensor([0, 0, 1, 1, 2, 2]), tensor([1, 2, 0, 1, 2, 0]), tensor([0, 1, 2, 3, 4, 5]))
+    >>> sg = dgl.sampling.sample_labors(g, [0, 1], 3, exclude_edges=[0, 1, 2])
+    >>> sg.all_edges(form='all')
+    (tensor([2, 1]), tensor([0, 1]), tensor([0, 1]))
+    >>> sg.has_edges_between(g_edges[0][:3],g_edges[1][:3])
+    tensor([False, False, False])
+    >>> g = dgl.heterograph({
+    ...   ('drug', 'interacts', 'drug'): ([0, 0, 1, 1, 3, 2], [1, 2, 0, 1, 2, 0]),
+    ...   ('drug', 'interacts', 'gene'): ([0, 0, 1, 1, 2, 2], [1, 2, 0, 1, 2, 0]),
+    ...   ('drug', 'treats', 'disease'): ([0, 0, 1, 1, 2, 2], [1, 2, 0, 1, 2, 0])})
+    >>> g_edges = g.all_edges(form='all', etype=('drug', 'interacts', 'drug'))
+    (tensor([0, 0, 1, 1, 3, 2]), tensor([1, 2, 0, 1, 2, 0]), tensor([0, 1, 2, 3, 4, 5]))
+    >>> excluded_edges  = {('drug', 'interacts', 'drug'): g_edges[2][:3]}
+    >>> sg = dgl.sampling.sample_labors(g, {'drug':[0, 1]}, 3, exclude_edges=excluded_edges)
+    >>> sg.all_edges(form='all', etype=('drug', 'interacts', 'drug'))
+    (tensor([2, 1]), tensor([0, 1]), tensor([0, 1]))
+    >>> sg.has_edges_between(g_edges[0][:3],g_edges[1][:3],etype=('drug', 'interacts', 'drug'))
+    tensor([False, False, False])
+
+    """
+    if F.device_type(g.device) == 'cpu' and not g.is_pinned():
+        frontier, importances = _sample_labors(
+            g, nodes, fanout, random_seed, edge_dir=edge_dir, prob=prob,
+            importance_sampling=importance_sampling, copy_ndata=copy_ndata,
+            copy_edata=copy_edata, exclude_edges=exclude_edges)
+    else:
+        frontier, importances = _sample_labors(
+            g, nodes, fanout, random_seed, edge_dir=edge_dir, prob=prob,
+            importance_sampling=importance_sampling, copy_ndata=copy_ndata, copy_edata=copy_edata)
+        if exclude_edges is not None:
+            eid_excluder = EidExcluder(exclude_edges)
+            if importance_sampling != 0:
+                frontier, importances = eid_excluder(frontier, importances)
+            else:
+                frontier = eid_excluder(frontier)
```

Comment:
I think this logic is common with `sample_neighbors`.  Is it possible to have a utility function to wrap this up?

Reply:
How can I handle the difference in the number of return returns values? Labor returns edge weights but sample_neighbors doesn't.

Reply:
Where could such a utility function be placed?

Reply:
`dgl/sampling/utils.py` could be one.

Reply:
Arguments can be handled by `(*args, **kwargs)`.  Return values is a problem though.  Let's table this for now.